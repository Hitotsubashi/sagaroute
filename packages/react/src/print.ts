import path from 'path';
import fs from 'fs';
import {
  INJECT_SYMBOL_PREFIX_REGEXP,
  TEMPLATE_RENDER_IMPORTS_SYMBOL,
  TEMPLATE_RENDER_ROUTE_SYMBOL,
} from '@/utils/symbol';
import { Imports, RouteObject } from '@/weave';
import { ParseResult } from '@babel/parser';
import traverse from '@babel/traverse';
import * as t from '@babel/types';
import generate from '@babel/generator';
import Mustache from 'mustache';
import hookCompose from '@/utils/hookCompose';
import parseToAst from '@/utils/parseToAst';
import transformRoutesToString from './utils/transformRoutesToString';

export type PrintHookBeforeParse = (routeFilePath: string) => void | string | null;
export type PrintHookAfterParse = (template: string, routeFilePath: string) => void;
export type PrintHookBeforeInject = (
  view: Record<string, any>,
  template: string,
  routes: RouteObject[],
) => void | string | null;
export type PrintHookAfterInject = (
  content: string,
  view: Record<string, any>,
  template: string,
) => void;
export type PrintHookBeforeWrite = (renderedContent: string, routeFilePath: string) => void | null;
export type PrintHookAfterWrite = (routeFilePath: string) => void;

interface PrintOption {
  routeFilePath?: string;
  onWarning?: (message: string) => void;
  hooks?: {
    parse?: {
      before?: PrintHookBeforeParse[];
      after?: PrintHookAfterParse[];
    };
    inject?: {
      before?: PrintHookBeforeInject[];
      after?: PrintHookAfterInject[];
    };
    write?: {
      before?: PrintHookBeforeWrite[];
      after?: PrintHookAfterWrite[];
    };
  };
}

function getRouteFileContent(routeFilePath: string): string {
  const stats = fs.statSync(routeFilePath, { throwIfNoEntry: false });
  if (!stats || !stats.isFile()) {
    throw new Error(`Could not find the file with path ${routeFilePath}`);
  }
  return fs.readFileSync(routeFilePath, 'utf-8');
}

const IMPORT_DECLARATIONS_COMMENT = 'injected by sagaroute';
export const IMPORT_DECLARATIONS_COMMENT_START = `${IMPORT_DECLARATIONS_COMMENT}: start`;
export const IMPORT_DECLARATIONS_COMMENT_END = `${IMPORT_DECLARATIONS_COMMENT}: end`;

function removeInjectedContent(content: string) {
  return content.replace(
    /\/\* injected by sagaroute: start \*\/.*\/\* injected by sagaroute: end \*\//gs,
    '',
  );
}

function parseStickyInjectCommentToPlaceholder(
  ast: ParseResult<t.File>,
  handledInjectVariables: string[],
) {
  function getInjectedVariable(node: t.Node) {
    const startLine = node.loc?.start.line;
    const { leadingComments } = node;
    const matchObj = leadingComments
      ?.find(({ loc }) => loc?.start && loc.start.line + 1 === startLine)
      ?.value.match(INJECT_SYMBOL_PREFIX_REGEXP);
    return matchObj?.[1]?.trim();
  }

  traverse(ast, {
    VariableDeclaration({ node }) {
      const startLine = node.loc?.start.line;
      const varName = getInjectedVariable(node);
      if (varName) {
        for (let i = 0; i < node.declarations.length; i++) {
          const declarator = node.declarations[i];
          if (
            t.isVariableDeclarator(declarator) &&
            declarator.loc &&
            /**
             * 防止一条声明语句里被拆分成多行的情况，如：
             * let code=x,
             *  code1=x1
             */
            declarator.loc.start.line === startLine
          ) {
            handledInjectVariables.push(varName);
            declarator.init = t.identifier(`{{{${varName}}}}`);
          }
        }
      }
    },
    ExportNamedDeclaration({ node }) {
      const startLine = node.loc?.start.line;
      const varName = getInjectedVariable(node);
      if (varName && t.isVariableDeclaration(node.declaration)) {
        for (let i = 0; i < node.declaration.declarations.length; i++) {
          const declarator = node.declaration.declarations[i];
          if (
            t.isVariableDeclarator(declarator) &&
            declarator.loc &&
            /**
             * 防止一条声明语句里被拆分成多行的情况，如：
             * let code=x,
             *  code1=x1
             */
            declarator.loc.start.line === startLine
          ) {
            handledInjectVariables.push(varName);
            declarator.init = t.identifier(`{{{${varName}}}}`);
          }
        }
      }
    },
    ReturnStatement({ node }) {
      const varName = getInjectedVariable(node);
      if (varName) {
        handledInjectVariables.push(varName);
        node.argument = t.identifier(`{{{${varName}}}}`);
      }
    },
    ExportDefaultDeclaration({ node }) {
      const varName = getInjectedVariable(node);
      if (varName) {
        handledInjectVariables.push(varName);
        node.declaration = t.identifier(`{{{${varName}}}}`);
      }
    },
  });
}

function parseEmptyInjectCommentToPlaceholder(content: string, handledInjectVariables: string[]) {
  const injectedVariables = new Set(handledInjectVariables);
  function replacer(match: string, p1: string) {
    const matchObj = p1.match(INJECT_SYMBOL_PREFIX_REGEXP);
    if (matchObj) {
      const varName = matchObj[1]?.trim();
      if (!handledInjectVariables.includes(varName)) {
        injectedVariables.add(varName);
        return [
          match,
          `/* ${IMPORT_DECLARATIONS_COMMENT_START} */`,
          `{{{${varName}}}}`,
          `/* ${IMPORT_DECLARATIONS_COMMENT_END} */`,
        ].join('\n');
      }
    }
    return match;
  }

  let modifiedContent = content.replace(/\/\/(.*)/g, replacer);
  modifiedContent = modifiedContent.replace(/\/\*(.*)\*\//g, replacer);
  return {
    template: modifiedContent,
    injectedVariables: Array.from(injectedVariables),
  };
}

function generateRenderTemplateFromContent(content: string, routeFilePath: string) {
  const handledInjectVariables: string[] = [];
  const modifiedContent = removeInjectedContent(content);
  const ast = parseToAst(
    modifiedContent,
    ['.ts', '.tsx'].includes(path.extname(routeFilePath)),
    routeFilePath,
  );
  const existedImportNames = collectExistedImportNames(ast);
  parseStickyInjectCommentToPlaceholder(ast, handledInjectVariables);
  let { code: template } = generate(ast);
  let injectedVariables: string[];
  // eslint-disable-next-line prefer-const
  ({ template, injectedVariables } = parseEmptyInjectCommentToPlaceholder(
    template,
    handledInjectVariables,
  ));
  return { template, existedImportNames, injectedVariables };
}

function collectExistedImportNames(ast: ParseResult<t.File>) {
  const existedImportNames: string[] = [];
  traverse(ast, {
    ImportDeclaration({ node }) {
      node.specifiers.forEach((specifier) => {
        if (
          t.isImportDefaultSpecifier(specifier) ||
          t.isImportNamespaceSpecifier(specifier) ||
          t.isImportSpecifier(specifier)
        ) {
          existedImportNames.push(specifier.local.name);
        }
      });
    },
  });
  return existedImportNames;
}

function transformImportsToString(imports: Imports, existedImportNames: string[]) {
  const importLines: string[] = [];

  Object.entries(imports).forEach(([source, dependencies]) => {
    const notDefaultImport: string[] = [];
    let defaultImport: string = '';
    dependencies
      .filter(({ asName }) => !existedImportNames.includes(asName))
      .forEach((dependency) => {
        if (dependency.isDefault) {
          // import x1 from 'xxx'
          if (dependency.name !== '*') {
            defaultImport = dependency.asName;
            // import * as x1 from 'xxx'
          } else {
            importLines.push(`import * as ${dependency.asName} from "${source}";`);
          }
        } else {
          // import {x1} from 'xxx'
          if (dependency.name === dependency.asName) {
            notDefaultImport.push(dependency.name);
          } else {
            notDefaultImport.push(`${dependency.name} as ${dependency.asName}`);
          }
        }
      });
    const importElementString = [
      defaultImport,
      notDefaultImport.length ? `{${notDefaultImport.join(',')}}` : '',
    ]
      .filter((item) => Boolean(item))
      .join(',');
    if (importElementString) {
      importLines.push(`import ${importElementString} from "${source}";`);
    }
  });
  return importLines.join('\n');
}

function diffVariableBetweenViewAndTemplate(
  viewVariable: string[],
  templateVariable: string[],
  onWarning: NonNullable<PrintOption['onWarning']>,
  routeFilePath: string,
) {
  const variablesOnlyInView = viewVariable.filter((item) => !templateVariable.includes(item));
  const variablesOnlyInTemplate = templateVariable.filter((item) => !viewVariable.includes(item));
  if (variablesOnlyInView.length) {
    onWarning(
      `The RoutingTemplateFile<${routeFilePath}> has no variables such as ${variablesOnlyInView
        .map((item) => `[${item}]`)
        .join(',')} which are needed in view.`,
    );
  }
  if (variablesOnlyInTemplate.length) {
    onWarning(
      `The view of stage<print.inject.before> has no variables such as ${variablesOnlyInTemplate
        .map((item) => `[${item}]`)
        .join(',')} which are needed in template.`,
    );
  }
}

export default function print(routes: RouteObject[], imports: Imports, option?: PrintOption) {
  const { routeFilePath = path.join('src', 'route.ts'), hooks, onWarning } = option ?? {};
  let renderTemplate = hookCompose(hooks?.parse?.before, routeFilePath);
  if (renderTemplate === null) return;
  let existedImportNames: string[] = [];
  let injectedVariables: undefined | string[];
  if (!renderTemplate) {
    const content = getRouteFileContent(routeFilePath);
    ({
      template: renderTemplate,
      existedImportNames,
      injectedVariables,
    } = generateRenderTemplateFromContent(content, routeFilePath));
  }
  hookCompose(hooks?.parse?.after, renderTemplate, routeFilePath);
  const view: Record<string, any> = {
    [TEMPLATE_RENDER_ROUTE_SYMBOL]: transformRoutesToString(routes),
    [TEMPLATE_RENDER_IMPORTS_SYMBOL]: transformImportsToString(imports, existedImportNames),
  };
  let renderedContent = hookCompose(hooks?.inject?.before, view, renderTemplate, routes);
  if (renderedContent === null) return;
  // 如果用户在parse.before中返回了template，此时injectedVariables为undefined
  if (injectedVariables && onWarning) {
    diffVariableBetweenViewAndTemplate(
      Object.keys(view),
      injectedVariables,
      onWarning,
      routeFilePath,
    );
  }
  if (!renderedContent) {
    renderedContent = Mustache.render(renderTemplate, view);
    renderedContent = renderedContent.replace(
      /\/\* injected by sagaroute: start \*\/\s*\/\* injected by sagaroute: end \*\/\s/gs,
      '',
    );
  }
  hookCompose(hooks?.inject?.after, renderedContent, view, renderTemplate);
  if (hookCompose(hooks?.write?.before, renderedContent, routeFilePath) !== null) {
    fs.writeFileSync(routeFilePath, renderedContent, {
      flag: 'w+',
    });
    hookCompose(hooks?.write?.after, routeFilePath);
  }
}
