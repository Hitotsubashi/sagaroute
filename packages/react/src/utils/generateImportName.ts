import path from 'path';

function capitalize(word: string) {
  return word.charAt(0).toUpperCase() + word.slice(1);
}

export default function generateImportName(fpath: string, varName: string): string {
  const { name, dir } = path.parse(fpath);
  let upper = true;
  let importName = '';
  let basepath = path.join(dir, name);
  for (let i = 0; i < basepath.length; i++) {
    // if (basepath[i] === ".") {
    //   upper = false;
    // } else if ([path.sep, "-", "_", "[", "]", "$"].includes(basepath[i])) {
    if (!/[A-Za-z0-9]/.test(basepath[i])) {
      upper = true;
    } else {
      if (upper) {
        importName += basepath[i].toUpperCase();
        upper = false;
      } else {
        importName += basepath[i];
      }
    }
  }
  if (!/[_A-Za-z]/.test(importName[0])) {
    importName = '_' + importName;
  }
  if (name === '[]') {
    importName += '_';
  }
  return importName + capitalize(varName);
}
