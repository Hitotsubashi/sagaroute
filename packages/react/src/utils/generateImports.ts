import { Dependency } from '@/gather';
import { Imports } from '@/weave';

export default function generateImports(dependencies: Dependency[] | undefined): Imports {
  const imports: Imports = {};
  dependencies?.forEach((dependency) => {
    const { importPath, ...rest } = dependency;
    if (imports[importPath]) {
      imports[importPath].push(rest);
    } else {
      imports[importPath] = [rest];
    }
  });
  return imports;
}
