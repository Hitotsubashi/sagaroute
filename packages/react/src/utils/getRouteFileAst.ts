import path from 'path';
import fs from 'fs';
import parseToAst from './parseToAst';

export function getRouteFilePath(filepath: string) {
  const { dir, ext, name } = path.parse(filepath);
  return dir + `${name}.route` + ext;
}

export function getRouteFileAstIfExist(routeFilePath: string) {
  const stats = fs.statSync(routeFilePath, { throwIfNoEntry: false });
  if (stats) {
    const content = fs.readFileSync(routeFilePath, 'utf-8');
    const isTsx = routeFilePath.endsWith('.tsx');
    const ast = parseToAst(content, isTsx, routeFilePath);
    return ast;
  }
  return null;
}
