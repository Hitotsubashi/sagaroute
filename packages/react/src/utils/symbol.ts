export const EVAL_STRING_SYMBOL = '/*#__EVAL__*/';
export const EVAL_STRING_SYMBOL_REGEXP = /"\/\*#__EVAL__\*\/(.*?)\/\*#__EVAL__\*\/"/g;
export const ROUTE_CONTENT_SYMBOL_REGEXP = /SAGA_ROUTE:?(.*)?/;
export const TEMPLATE_RENDER_ROUTE_SYMBOL = 'routes';
export const TEMPLATE_RENDER_IMPORTS_SYMBOL = 'imports';
export const INJECT_SYMBOL_PREFIX = 'sagaroute-inject';
export const INJECT_SYMBOL_PREFIX_REGEXP = /sagaroute-inject:(.*)/;

export const htmlTagNames = [
  'a',
  'abbr',
  'acronym',
  'address',
  'applet',
  'area',
  'article',
  'aside',
  'audio',
  'b',
  'base',
  'basefont',
  'bdi',
  'bdo',
  'bgsound',
  'big',
  'blink',
  'blockquote',
  'body',
  'br',
  'button',
  'canvas',
  'caption',
  'center',
  'cite',
  'code',
  'col',
  'colgroup',
  'command',
  'content',
  'data',
  'datalist',
  'dd',
  'del',
  'details',
  'dfn',
  'dialog',
  'dir',
  'div',
  'dl',
  'dt',
  'element',
  'em',
  'embed',
  'fieldset',
  'figcaption',
  'figure',
  'font',
  'footer',
  'form',
  'frame',
  'frameset',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'head',
  'header',
  'hgroup',
  'hr',
  'html',
  'i',
  'iframe',
  'image',
  'img',
  'input',
  'ins',
  'isindex',
  'kbd',
  'keygen',
  'label',
  'legend',
  'li',
  'link',
  'listing',
  'main',
  'map',
  'mark',
  'marquee',
  'math',
  'menu',
  'menuitem',
  'meta',
  'meter',
  'multicol',
  'nav',
  'nextid',
  'nobr',
  'noembed',
  'noframes',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'picture',
  'plaintext',
  'pre',
  'progress',
  'q',
  'rb',
  'rbc',
  'rp',
  'rt',
  'rtc',
  'ruby',
  's',
  'samp',
  'script',
  'search',
  'section',
  'select',
  'shadow',
  'slot',
  'small',
  'source',
  'spacer',
  'span',
  'strike',
  'strong',
  'style',
  'sub',
  'summary',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'template',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'track',
  'tt',
  'u',
  'ul',
  'var',
  'video',
  'wbr',
  'xmp',
];

export const svgTagNames = [
  'a',
  'altGlyph',
  'altGlyphDef',
  'altGlyphItem',
  'animate',
  'animateColor',
  'animateMotion',
  'animateTransform',
  'animation',
  'audio',
  'canvas',
  'circle',
  'clipPath',
  'color-profile',
  'cursor',
  'defs',
  'desc',
  'discard',
  'ellipse',
  'feBlend',
  'feColorMatrix',
  'feComponentTransfer',
  'feComposite',
  'feConvolveMatrix',
  'feDiffuseLighting',
  'feDisplacementMap',
  'feDistantLight',
  'feDropShadow',
  'feFlood',
  'feFuncA',
  'feFuncB',
  'feFuncG',
  'feFuncR',
  'feGaussianBlur',
  'feImage',
  'feMerge',
  'feMergeNode',
  'feMorphology',
  'feOffset',
  'fePointLight',
  'feSpecularLighting',
  'feSpotLight',
  'feTile',
  'feTurbulence',
  'filter',
  'font',
  'font-face',
  'font-face-format',
  'font-face-name',
  'font-face-src',
  'font-face-uri',
  'foreignObject',
  'g',
  'glyph',
  'glyphRef',
  'handler',
  'hkern',
  'iframe',
  'image',
  'line',
  'linearGradient',
  'listener',
  'marker',
  'mask',
  'metadata',
  'missing-glyph',
  'mpath',
  'path',
  'pattern',
  'polygon',
  'polyline',
  'prefetch',
  'radialGradient',
  'rect',
  'script',
  'set',
  'solidColor',
  'stop',
  'style',
  'svg',
  'switch',
  'symbol',
  'tbreak',
  'text',
  'textArea',
  'textPath',
  'title',
  'tref',
  'tspan',
  'unknown',
  'use',
  'video',
  'view',
  'vkern',
];
