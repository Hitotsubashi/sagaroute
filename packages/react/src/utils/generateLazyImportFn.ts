import { FileNode } from "@/gather";
import { EVAL_STRING_SYMBOL } from "./symbol";
import { Imports, RouteObject } from "@/weave";
import generateImports from "./generateImports";
import mergeImports from "./mergeImports";
import transformRoutesToString from "./transformRoutesToString";
import normalizePath from "./normalizePath";
import generateImportName from "./generateImportName";

function generateImportLines(imports: Imports): string {
  const importLines: string[] = [];

  Object.entries(imports).forEach(([source, dependencies]) => {
    const notDefaultImport: string[] = [];
    let defaultImport: string = "";
    dependencies.forEach((dependency) => {
      if (dependency.isDefault) {
        // import x1 from 'xxx'
        if (dependency.name !== "*") {
          notDefaultImport.push(`default: ${dependency.asName}`);
          // import * as x1 from 'xxx'
        } else {
          importLines.push(
            `const ${dependency.asName} = await import("${source}");`
          );
        }
      } else {
        // import {x1} from 'xxx'
        if (dependency.name === dependency.asName) {
          notDefaultImport.push(dependency.name);
        } else {
          notDefaultImport.push(`${dependency.name} : ${dependency.asName}`);
        }
      }
    });
    const importElementString = [
      defaultImport,
      notDefaultImport.length ? `{${notDefaultImport.join(",")}}` : "",
    ]
      .filter((item) => Boolean(item))
      .join(",");
    if (importElementString) {
      importLines.push(
        `const ${importElementString}  = await import("${source}");`
      );
    }
  });
  return importLines.join("\n");
}

interface Option {
  pathRewrite?: [RegExp, string][];
  restRouteProps: Omit<
    RouteObject,
    "path" | "index" | "children" | "caseSensitive"
  >;
  dependencies: FileNode["dependencies"];
}

export default function generateLazyImportFn(
  componentPath: string,
  { restRouteProps, dependencies, pathRewrite }: Option
) {
  const asName = generateImportName(componentPath, "");
  const adjustedPath = normalizePath(componentPath, pathRewrite);
  const imports = generateImports(dependencies);
  mergeImports(imports, {
    [adjustedPath]: [
      {
        name: "default",
        asName,
        isDefault: false,
      },
    ],
  });
  const importLines = generateImportLines(imports);
  const returnObj: RouteObject = {
    Component: `${EVAL_STRING_SYMBOL}${asName}`,
    ...restRouteProps,
  };
  return `${EVAL_STRING_SYMBOL}async function() {
    ${importLines}
    return ${transformRoutesToString(returnObj)};
  }`;
}
