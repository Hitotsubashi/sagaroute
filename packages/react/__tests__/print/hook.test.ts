import print from "@/print";
import { EVAL_STRING_SYMBOL } from "@/utils/symbol";
import path from "path";
import fs from "fs";
import Mustache from "mustache";
import { Imports, RouteObject } from "@/weave";
import { transformRoutesToString } from "@/index";

test("test print with hooks", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      children: [
        {
          path: ":id",
          element: `${EVAL_STRING_SYMBOL}<SrcPagesUsers_Id_/>`,
        },
      ],
    },
    {
      path: "/",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesIndex/>`,
    },
    {
      path: "*",
      element: `${EVAL_STRING_SYMBOL}<SrcPages404/>`,
    },
    {
      path: "permission",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesPermissionIndex/>`,
    },
  ];
  const imports: Imports = {
    "./src/pages/users/[id].tsx": [
      {
        name: "User",
        asName: "SrcPagesUsers_Id_",
        isDefault: true,
      },
    ],
    "./src/pages/index.tsx": [
      {
        name: "Index",
        asName: "SrcPagesIndex",
        isDefault: true,
      },
    ],
    "./src/pages/404.tsx": [
      {
        name: "Page404",
        asName: "SrcPages404",
        isDefault: true,
      },
    ],
    "./src/pages/permission/index.tsx": [
      {
        name: "Permission",
        asName: "SrcPagesPermissionIndex",
        isDefault: true,
      },
    ],
  };
  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route3.tsx"
  );

  const printHookBeforeParse = jest.fn((routeFilePath: string) => {});
  const printHookAfterParse = jest.fn(
    (template: string, routeFilePath: string) => {}
  );
  const printHookBeforeInject = jest.fn(
    (view: Record<string, any>, template: string, routes: RouteObject[]) => {
      view["constantRoutes"] = transformRoutesToString(
        routes.filter((item: any) => item.path !== "permission")
      );
      view["asyncRoutes"] = transformRoutesToString(
        routes.filter((item: any) => item.path === "permission")
      );
      view["prefix"] = `import React from "react";`;
    }
  );
  const printHookAfterInject = jest.fn(() => {});

  print(routes, imports, {
    routeFilePath,
    hooks: {
      parse: {
        before: [printHookBeforeParse],
        after: [printHookAfterParse],
      },
      inject: {
        before: [printHookBeforeInject],
        after: [printHookAfterInject],
      },
    },
  });
  expect(printHookBeforeParse.mock.calls).toHaveLength(1);
  expect(printHookBeforeParse.mock.calls[0]).toStrictEqual([routeFilePath]);
  expect(printHookAfterParse.mock.calls).toHaveLength(1);
  expect(printHookAfterParse.mock.calls[0][1]).toBe(routeFilePath);
  expect(printHookBeforeInject.mock.calls).toHaveLength(1);
  expect(printHookAfterInject.mock.calls).toHaveLength(1);

  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});

test("test print with hook using custom template", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      children: [
        {
          path: ":id",
          element: `${EVAL_STRING_SYMBOL}<SrcPagesUsers_Id_/>`,
        },
      ],
    },
    {
      path: "/",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesIndex/>`,
    },
    {
      path: "*",
      element: `${EVAL_STRING_SYMBOL}<SrcPages404/>`,
    },
  ];
  const imports: Imports = {
    "./src/pages/users/[id].tsx": [
      {
        name: "User",
        asName: "SrcPagesUsers_Id_",
        isDefault: true,
      },
    ],
    "./src/pages/index.tsx": [
      {
        name: "Index",
        asName: "SrcPagesIndex",
        isDefault: true,
      },
    ],
    "./src/pages/404.tsx": [
      {
        name: "Page404",
        asName: "SrcPages404",
        isDefault: true,
      },
    ],
  };
  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route5.tsx"
  );
  print(routes, imports, {
    routeFilePath,
    hooks: {
      parse: {
        before: [
          () => {
            return fs.readFileSync(
              path.join(__dirname, "ignore-test-render", "route5.tpl"),
              "utf-8"
            );
          },
        ],
      },
    },
  });
  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});

test("test print with hook skip render.", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      children: [
        {
          path: ":id",
          element: `${EVAL_STRING_SYMBOL}<SrcPagesUsers_Id_/>`,
        },
      ],
    },
    {
      path: "/",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesIndex/>`,
    },
    {
      path: "*",
      element: `${EVAL_STRING_SYMBOL}<SrcPages404/>`,
    },
  ];
  const imports: Imports = {
    "./src/pages/users/[id].tsx": [
      {
        name: "User",
        asName: "SrcPagesUsers_Id_",
        isDefault: true,
      },
    ],
    "./src/pages/index.tsx": [
      {
        name: "Index",
        asName: "SrcPagesIndex",
        isDefault: true,
      },
    ],
    "./src/pages/404.tsx": [
      {
        name: "Page404",
        asName: "SrcPages404",
        isDefault: true,
      },
    ],
  };
  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route6.tsx"
  );
  print(routes, imports, {
    routeFilePath,
    hooks: {
      inject: {
        before: [
          (view, template) => {
            return Mustache.render(template, {
              routes: 1,
              imports: ['import xx from "xxx";'].join("\n"),
            });
          },
        ],
      },
    },
  });
  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});

test("test print with hook skip write.", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      children: [
        {
          path: ":id",
          element: `${EVAL_STRING_SYMBOL}<SrcPagesUsers_Id_/>`,
        },
      ],
    },
    {
      path: "/",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesIndex/>`,
    },
    {
      path: "*",
      element: `${EVAL_STRING_SYMBOL}<SrcPages404/>`,
    },
  ];
  const imports: Imports = {
    "./src/pages/users/[id].tsx": [
      {
        name: "User",
        asName: "SrcPagesUsers_Id_",
        isDefault: true,
      },
    ],
    "./src/pages/index.tsx": [
      {
        name: "Index",
        asName: "SrcPagesIndex",
        isDefault: true,
      },
    ],
    "./src/pages/404.tsx": [
      {
        name: "Page404",
        asName: "SrcPages404",
        isDefault: true,
      },
    ],
  };
  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route7.tsx"
  );
  let renderContentRecord = "";
  print(routes, imports, {
    routeFilePath,
    hooks: {
      write: {
        before: [
          (renderedContent: string) => {
            renderContentRecord = renderedContent;
            return null;
          },
        ],
      },
    },
  });
  expect(renderContentRecord).toMatchSnapshot();
  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});
