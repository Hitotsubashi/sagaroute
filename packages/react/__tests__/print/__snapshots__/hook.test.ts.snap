// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test print with hook skip render. 1`] = `
"// @ts-nocheck
import xx1 from "xxx";
/* sagaroute-inject:imports */
/* injected by sagaroute: start */
import xx from "xxx";
/* injected by sagaroute: end */

// sagaroute-inject: routes
const routes = 1;
export default null;"
`;

exports[`test print with hook skip write. 1`] = `
"// @ts-nocheck
import React from "react";
/* sagaroute-inject:imports */
/* injected by sagaroute: start */
import SrcPagesUsers_Id_ from "./src/pages/users/[id].tsx";
import SrcPagesIndex from "./src/pages/index.tsx";
import SrcPages404 from "./src/pages/404.tsx";
/* injected by sagaroute: end */

// sagaroute-inject: routes
const routes = [
  {
    "path": "users",
    "children": [
      {
        "path": ":id",
        "element": <SrcPagesUsers_Id_/>
      }
    ]
  },
  {
    "path": "/",
    "element": <SrcPagesIndex/>
  },
  {
    "path": "*",
    "element": <SrcPages404/>
  }
];
export default routes;"
`;

exports[`test print with hook skip write. 2`] = `
"// @ts-nocheck
import React from "react";
/* sagaroute-inject:imports */

// sagaroute-inject: routes
const routes = [];

export default routes;
"
`;

exports[`test print with hook using custom template 1`] = `
"// @ts-nocheck
import React from "react";
import SrcPagesUsers_Id_ from "./src/pages/users/[id].tsx";
import SrcPagesIndex from "./src/pages/index.tsx";
import SrcPages404 from "./src/pages/404.tsx";

const routes = [
  {
    "path": "users",
    "children": [
      {
        "path": ":id",
        "element": <SrcPagesUsers_Id_/>
      }
    ]
  },
  {
    "path": "/",
    "element": <SrcPagesIndex/>
  },
  {
    "path": "*",
    "element": <SrcPages404/>
  }
];

const router = createBrowserRouter(routes)

export default router;
"
`;

exports[`test print with hooks 1`] = `
"// @ts-nocheck
/* sagaroute-inject: prefix */
/* injected by sagaroute: start */
import React from "react";
/* injected by sagaroute: end */

// sagaroute-inject: constantRoutes
export const constantRoutes = [
  {
    "path": "users",
    "children": [
      {
        "path": ":id",
        "element": <SrcPagesUsers_Id_/>
      }
    ]
  },
  {
    "path": "/",
    "element": <SrcPagesIndex/>
  },
  {
    "path": "*",
    "element": <SrcPages404/>
  }
];

/* sagaroute-inject: asyncRoutes */
export const asyncRoutes = [
  {
    "path": "permission",
    "element": <SrcPagesPermissionIndex/>
  }
];"
`;
